src/microceph_adopt_ceph.py

```diff
@@ -0,0 +1,177 @@
+        if self.charm.ready_for_service():
+            logger.debug("Not processing adopt relation event, microceph already bootstrapped.")
+            return
+
+        logger.debug("Emitting adopt-ceph reconcile event")
+        self.on.adopt_ceph_bootstrap.emit(event.relation)
+
+    def _on_relation_broken(self, event) -> None:
+        """On relation departed."""
+        if not self.model.unit.is_leader():
+            logger.debug("Unit is not leader, skipping adopt-ceph departed event")
+            return
+
+        with sunbeam_guard.guard(self.charm, self.relation_name):
+            if not self.charm.ready_for_service():
+                raise sunbeam_guard.BlockedExceptionError(
+                    "Adopt relation removed before cluster bootstrap could be performed"
+                )
+
+        logger.debug("Ignoring adopt-ceph departed event")
+
+
+class AdoptCephRequiresHandler(RelationHandler):
+    """Handler for adopt-ceph relation events."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str,
+        callback_f: Callable,
+    ):
+        super().__init__(charm, relation_name, callback_f)
+
+    @property
+    def ready(self) -> bool:
+        """Check if adopt-ceph relation is ready."""
+        logger.info(f"Report {self.relation_name} as ready")
+        return True
+
+    def setup_event_handler(self) -> Object:
+        """Configure event handlers for a ceph-admin interface."""
+        logger.debug("Setting up adopt-ceph event handler")
+
+        self.adopt_ceph = AdoptCephRequires(self.charm, self.relation_name)
+
+        self.framework.observe(self.adopt_ceph.on.adopt_ceph_bootstrap, self._on_bootstrap)
+
+    def _on_bootstrap(self, relation):
+        """Bootstrap MicroCeph cluster using adopted ceph cluster."""
+        logger.info("Handling adopt-ceph bootstrap event")
+        with sunbeam_guard.guard(self.charm, self.relation_name):
+            for relation in self.model.relations.get(self.relation_name, []):
+                if not relation.units:
+                    logger.debug("No units in adopt-ceph relation, cannot reconcile")
+                    return
+
+                unit = next(iter(relation.units), None)
+                if unit is None:
+                    logger.debug("No units available in adopt-ceph relation after check")
+                    return
+                remote_ceph_data = relation.data.get(unit, {})
+                logger.debug(f"Adopt-ceph relation data: IsEmpty({remote_ceph_data is None})")
+
+                # fetched mon hosts value is a space separated string of host addresses.
+                mon_hosts = remote_ceph_data.get(AdoptCephRelationDataKeys.mon_hosts.value, None)
+                fsid = remote_ceph_data.get(AdoptCephRelationDataKeys.fsid.value, None)
+                admin_key = remote_ceph_data.get(AdoptCephRelationDataKeys.admin_key.value, None)
+
+                logger.debug(
```

Comment:
Nit: important lifecycle info, would make this logger.info() (no pun intended)

Reply:
ack